Add Clip enum to Material and related widgets

liyuqian opened the pull request:
See details in [our proposal for this breaking API change](https://groups.google.com/forum/#!searchin/flutter-dev/liyuqian@google.com%7Csort:date/flutter-dev/XMHE0XdsXxI/nonrNQpbBAAJ) and https://github.com/flutter/flutter/issues/18057

liyuqian committed:
Squash changes of Clip enum so far

Add Clip enum

Propogate Clip for Material in dart

The Clip isn't sent to the C++ engine yet, and it currently
only works for needsCompositing render objects.

liyuqian commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1020,6 +1020,9 @@ abstract class CustomClipper<T> {
   T getClip(Size size);
 
+  /// How to clip.
+  ClipOption get clipOption => ClipOption.antiAlias;
```
I'm not sure whether I should default to ClipOption.none here. It seems ironical to have no clip as default for a clipper. Currently, only ShapeBorderClipper default to ClipOption.none.

Hixie replied:
yeah the clipping widgets should probably default to `Clip.antiAlias`

liyuqian replied:
Ack.

liyuqian commented:
@Hixie @goderbauer @gspencergoog : one week has passed since we posted the breaking API change proposal so I'm starting to implement the change. This is an incomplete PR but I'd like to get some feedback before I go too far. Please let me know if I'm heading the right direction in dart.

Hixie commented on packages/flutter/lib/src/painting/basic_types.dart:
```diff
@@ -269,3 +269,27 @@ bool axisDirectionIsReversed(AxisDirection axisDirection) {
+  /// cases, consider whether you can avoid overlaying multiple colors in one
+  /// spot (e.g. by having the background color only present where the image is
+  /// absent).
+  antiAliasWithSaveLayer
```
for these, we should follow the following convention (used elsewhere in the docs):

```
/// Brief sentence fragment that says what it does.
///
/// This is a longer sentence that describes the semantics of the value in detail,
/// including edge cases. There might be more sentences here.
///
/// Then, have a sentence or more than talks about the current performance
/// implications.
///
/// Finally, a brief comment if the feature is unlikely to be used.
///
/// See also:
///
///  * [SomethingElse], which is something relevant.
```

The "see also" block would be omitted if not useful. But e.g. `antiAlias` might point to `Canvas.antiAlias`.

So for example:

```
/// Clip, but do not apply anti-aliasing.
///
/// This mode enables clipping, but curves and non-axis-aligned straight lines will be
/// jagged as no effort is made to anti-alias.
///
/// Faster than other clipping modes, but slower than [none].
///
/// This is a reasonable choice when clipping is needed, if the container is an axis-
/// aligned rectangle or an axis-aligned rounded rectangle with very small corner radii.
///
/// See also:
///
///  * [antiAlias], which is more reasonable when clipping is needed and the shape is not
///    an axis-aligned rectangle.
hardEdge,
```

liyuqian replied:
Thank you @Hixie for the sample writing! However, I couldn't find `Canvas.antiAlias`. It seems that we don't offer non-anti-aliased draw in `Canvas`?

Hixie replied:
Ah it's on Paint: https://master-docs-flutter-io.firebaseapp.com/flutter/dart-ui/Paint/isAntiAlias.html

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -840,6 +844,7 @@ class PhysicalModelLayer extends ContainerLayer {
+    this.clip = ClipOption.none,
     @required this.elevation,
     @required this.color,
     @required this.shadowColor,
```
assert clip != null

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1548,6 +1560,9 @@ class RenderPhysicalModel extends _RenderPhysicalModelBase<RRect> {
     _markNeedsClip();
   }
 
+  /// Whether and how to clip.
```
by the way if you come up with detailed docs for these and want to avoid copying-and-pasting them all over the place, and if they do in fact apply equally everywhere, then consider using a dartdoc macro/template to avoid duplication. See the child macro (used for the `child` property of most widgets) as an example.

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/painting/basic_types.dart:
```diff
@@ -269,3 +269,27 @@ bool axisDirectionIsReversed(AxisDirection axisDirection) {
 }
+
+/// Different ways to clip for a widget
+enum ClipOption {
```
the word "option" here doesn't seem to add much. Is `Clip` ok?

liyuqian replied:
Yeah, that's Ok. Renaming now.

Hixie commented:
ideally this PR would remove most if not all uses of ClipRect, ClipRRect, and ClipPath from the framework too, or at least make them configurable with a `clip` property as with Material. Similarly, most widgets that today use a `Material` should expose a `clip` that they pass through (e.g. `Card`, `Dialog`, etc).

liyuqian commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -605,6 +615,9 @@ class _MaterialInterior extends ImplicitlyAnimatedWidget {
   /// determines the physical shape.
   final ShapeBorder shape;
 
+  // Whether and how to clip
```
Sure, writing now.

liyuqian commented:
@Hixie : I just did a quick search on `new Material` and it's shocking to me how many widgets are creating Material in their build functions! Do I have to add the clip option to all of them in this PR?

It will make this PR really big. The upside is that we can easily revert the change. The downside is that reviewing becomes quite challenging.

How about we first create a PR which default clip to Clip.antiAlias first, then gradually expose clip options to many widgets in following PRs, and then finally set the default to Clip.none?

liyuqian commented:
@Hixie never mind, just figured out that most usages are in `test/cupertino` and `test/material`. They're not actual widgets. Will add clip option to all widgets inside `src/material`.

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1063,18 +1066,26 @@ class ShapeBorderClipper extends CustomClipper<Path> {
   /// towards the start of the shape).
   final TextDirection textDirection;
 
+
```
nit: extraneous blank line

Hixie commented on packages/flutter/lib/src/painting/basic_types.dart:
```diff
@@ -269,3 +269,77 @@ bool axisDirectionIsReversed(AxisDirection axisDirection) {
   return null;
 }
+
+/// Different ways to clip a widget's content
```
nit: end sentences with a period

Hixie commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -226,6 +230,12 @@ class Material extends StatefulWidget {
 
+  /// {@template flutter.widgets.Clip}
+  /// The content will be clipped (or not) according to this option.
+  /// See enum [Clip] for details of all possible options and their common use cases.
```
put a blank line between the two paragraphs

Hixie commented:
It might make sense to separate it into two PRs, I don't feel strongly one way or the other. But if we do split it, we should make sure that the first PR doesn't actually change any behaviour, so that we have a single PR that does the breaking change across everything.

We want to look for more than just `new Material`, by the way. We want to look for anywhere that is using any of the clipping widgets, physical models, material, `saveLayer`, etc. Clean them all out at once.

liyuqian committed:
Some nits

liyuqian committed:
Add defaultClipBehavior

This way, we can implement the breaking change in two steps.
In the first step, `defaultClipBehavior = Clip.antiAlias` so nothing
should be changed in rendering. In step two, we set
`defaultClipBehavior = Clip.none` and update all necessary tests.

liyuqian committed:
Add clip to CustomClipper

We have to rename several fields and methods to avoid the name conflict.

liyuqian committed:
Temp check

liyuqian committed:
Rebase with the new engine

liyuqian committed:
Add missing offset shift

liyuqian committed:
Handle all TODOs of respecting clip

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1016,15 +1017,18 @@ abstract class CustomClipper<T> {
   /// Returns a description of the clip given that the render object being
   /// clipped is of the given size.
-  T getClip(Size size);
+  T getClipShape(Size size);
```
Renaming this is a breaking change.

Hixie replied:
The old name is unfortunate, for sure, but I'm not sure it's unfortunate enough to warrant the cost of breaking customers.

Hixie replied:
There's a bunch of privates that you had to rename to avoid conflicting with `clip` too. Maybe we should rename `clip` to `clipBehavior` or `clipRegime` or some such?

liyuqian replied:
In that case, maybe we shall just call it `clipMode` to be consistent with flutter/engine?

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1016,15 +1017,18 @@ abstract class CustomClipper<T> {
+  T getClipShape(Size size);
 
-  /// Returns an approximation of the clip returned by [getClip], as
+  /// How to clip.
```
use the macro

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -87,13 +89,15 @@ abstract class MaterialInkController {
+/// 2. Ink effects: Material shows ink effects implemented by [InkFeature]s
 ///    like [InkSplash] and [InkHighlight] below its children.
 ///
+/// By default, [clip] is [Clip.none] and no clipping is performed. You
```
when writing docs, avoid referring to the reader ("You may..."). The documentation uses the passive voice, because we don't know the context of the reader. For example, maybe the reader is trying to understand existing code they can't change (in which case, they can't override it).

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -341,27 +355,29 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
       ),
       clipper: new ShapeBorderClipper(
         shape: shape,
+        clip: clip
```
nit: always add trailing commas

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -670,11 +673,11 @@ class ClipRRectLayer extends ContainerLayer {
     bool enabled = true;
     assert(() {
-      enabled = !debugDisableClipLayers;
+      enabled = !debugDisableClipLayers && clip != Clip.none;
```
this code only runs in debug mode. Is that what you intended?

liyuqian replied:
Ah, no. Changed.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -925,10 +931,13 @@ class PhysicalModelLayer extends ContainerLayer {
     @required this.shadowColor,
   }) : assert(clipPath != null),
+       assert(clip != null),
+       assert(clip != null),
```
redundant

liyuqian replied:
Fixed.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -939,6 +948,9 @@ class PhysicalModelLayer extends ContainerLayer {
   /// (as described at [Layer]).
   Path clipPath;
 
+  /// Whether and how to clip its children.
```
"its" doesn't make sense here since there's no antecedent.

(I'd probably just use the macro.)

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -341,18 +342,13 @@ class PaintingContext {
   /// * `painter` is a callback that will paint with the `clipRRect` applied. This
   ///   function calls the `painter` synchronously.
-  void pushClipRRect(bool needsCompositing, Offset offset, Rect bounds, RRect clipRRect, PaintingContextCallback painter) {
+  void pushClipRRect(bool needsCompositing, Offset offset, Rect bounds, RRect clipRRect, Clip clip, PaintingContextCallback painter) {
```
this is a breaking change. We should add the clip argument as a named argument with a default value.

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -2605,6 +2596,41 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
+      canvas.restore();
+    }
+    canvas.restore();
+  }
```
This belongs in the painting layer.

I recommend splitting it out into three versions so you get better type checking (they can still all be implemented in terms of a private version that uses the closure like you do here).

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1067,8 +1071,9 @@ class ShapeBorderClipper extends CustomClipper<Path> {
+    Clip clip = defaultClipBehavior, // ignore: deprecated_member_use
     this.textDirection,
-  }) : assert(shape != null);
+  }) : assert(shape != null), assert(clip != null), super(clip: clip);
```
no need to check clip if the superclass is going to check it

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1016,15 +1017,18 @@ abstract class CustomClipper<T> {
   ///
   /// The clipper will update its clip whenever [reclip] notifies its listeners.
-  const CustomClipper({ Listenable reclip }) : _reclip = reclip;
+  const CustomClipper({ Listenable reclip, this.clip = Clip.antiAlias }) : _reclip = reclip;
```
assert clip != null

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1133,24 +1139,28 @@ abstract class _RenderCustomClip<T> extends RenderProxyBox {
+  T get _defaultClipShape;
+  T _clipShape;
+
+  final Clip initialClip; // May be overridden by clipper.
```
i don't understand the purpose of initialClip

liyuqian replied:
Renamed it to defaultClip. Its purpose is to serve when the clipper is null: `Clip get clip => clipper == null ? defaultClip : clipper.clip;`

Hixie replied:
ah. add docs explaining all this. :-)

Hixie replied:
also, this should be mutable, otherwise you won't be able to change it when you update the render object widget.

Hixie replied:
maybe it should just be called the same as all the other properties that do the same thing?

liyuqian replied:
Done. BTW, it seems that our current API doesn't have any way of modifying `_clipBehavior` (the old `_defaultClipBehavior` and `initalClip`) when the render object widget is updated? The `_clip` is updated according to size, but `_clipBehavior` seems to remain constant with different sizes.

liyuqian committed:
Fix according to various comments.

liyuqian committed:
Rename clip to clipBehavior to avoid name conflicts

Hixie commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -691,6 +693,9 @@ class PhysicalModel extends SingleChildRenderObjectWidget {
   final BoxShape shape;
 
+  /// {@macro flutter.widgets.Clip}
+  final Clip clip;
```
you also need to update the clippers, right? ClipRect, ClipRRect, etc.

liyuqian replied:
ClipRRect, ClipRect... are updated through CustomClipper<T> (which you last commented in proxy_box.dart)

liyuqian replied:
Done.

liyuqian committed:
Add clipBehavior to ClipPath, ClipRRect, ClipRect, ClipOval

liyuqian committed:
Update clip to clipBehavior after engine rename

liyuqian commented:
@Hixie PTAL as clip is now renamed to clipBehavior.

Hixie commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -87,13 +89,15 @@ abstract class MaterialInkController {
 ///    [elevation] pixels, and draws the appropriate shadow.
-/// 3. Ink effects: Material shows ink effects implemented by [InkFeature]s
+/// 2. Ink effects: Material shows ink effects implemented by [InkFeature]s
 ///    like [InkSplash] and [InkHighlight] below its children.
```
I think we should probably leave the list as it was, but make it clear the clipping is only if requested, just like the shadow is only if requested, and the ink effects are only if requested.

Hixie commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -226,6 +232,13 @@ class Material extends StatefulWidget {
+  /// {@template flutter.widgets.Clip}
+  /// The content will be clipped (or not) according to this option.
+  ///
+  /// See enum [Clip] for details of all possible options and their common use cases.
```
grammar: "See the [Clip] enum for..."

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -123,6 +123,56 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   }
+
+  static void _clipAndPaint<T>(Canvas canvas, Clip clipBehavior, T t, Rect bounds, void painter()) {
+    void Function(bool doAA) canvasClipCall;
```
nit: avoid abbrevations (AA)

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -123,6 +123,56 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// Clip [Canvas] with [Path] according to [Clip] and then paint. [Canvas] is
+  /// restored to the pre-clip status afterwards.
+  static void clipPathAndPaint(Canvas canvas, Clip clipBehavior, Path path, Rect bounds, void painter()) {
+    _clipAndPaint(canvas, clipBehavior, path, bounds, painter);
```
you can inline more of this. There's no point this code path going through and checking the type of `path`, for instance. Just pass in the canvasClipCall directly to `_clipAndPaint`.

liyuqian replied:
Did I misunderstand https://github.com/flutter/flutter/pull/18576#discussion_r204920136 ?

It was sent directly to _clipAndPaint but the comment above suggests breaking into 3 function calls to check the type.

liyuqian replied:
Got it after offline discussion. Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -605,14 +655,17 @@ class ClipRectLayer extends ContainerLayer {
   /// The [clipRect] property must be non-null before the compositing phase of
   /// the pipeline.
-  ClipRectLayer({ this.clipRect });
+  ClipRectLayer({ this.clipRect, this.clipBehavior = Clip.antiAlias }) : assert(clipBehavior != null && clipBehavior != Clip.none);
```
`assert(foo && bar)` is better written as `assert(foo), assert(bar)` because then when they fail, the error will be on the expression that failed, rather than one of two expressions that failed.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -651,14 +704,17 @@ class ClipRRectLayer extends ContainerLayer {
   /// The [clipRRect] property must be non-null before the compositing phase of
   /// the pipeline.
-  ClipRRectLayer({ this.clipRRect });
+  ClipRRectLayer({ this.clipRRect, this.clipBehavior = Clip.antiAlias });
```
assert?

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -697,14 +753,20 @@ class ClipPathLayer extends ContainerLayer {
   /// The [clipPath] property must be non-null before the compositing phase of
   /// the pipeline.
-  ClipPathLayer({ this.clipPath });
+  ClipPathLayer({ this.clipPath, this.clipBehavior = Clip.antiAlias }) : assert(clipBehavior != null && clipBehavior != Clip.none);
```
as above

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -697,14 +753,20 @@ class ClipPathLayer extends ContainerLayer {
+  /// {@template flutter.layer.clipBehavior}
+  /// Controls how to clip (default to [Clip.antiAlias]). [Clip.none] is not
+  /// allowed here.
+  /// {@endtemplate}
```
put the sentence starting with `[Clip.none]` in a separate paragraph.

Hixie replied:
(the first sentence is in its own paragraph because it's used in more places)

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -651,14 +704,17 @@ class ClipRRectLayer extends ContainerLayer {
   RRect clipRRect;
 
+  /// {@macro flutter.layer.clipBehavior}
+  Clip clipBehavior;
```
this says it's not allowed to be `Clip.none`, which seems reasonable, but you don't assert it above (which you do in other constructors) and you do support it in `addToScene` (which you don't in the others), which seems confusing. I think they should all be consistent (and probably not support none).

Hixie replied:
It seems weird that you support it in addToScene and indeed have a mutable field that can be set to `none` but don't allow it in the constructor. I hadn't realized it was a mutable field before (it makes sense that it is). Maybe the field should have a setter that asserts it's valid?

as in:

```dart
  Clip get clipBehavior => _clipBehavior;
  Clip _clipBehavior;
  set clipBehavior(Clip value) {
    assert(value != null);
    assert(value != Clip.none);
    _clipBehavior = value;
  }
```

Hixie replied:
if you did that you could remove the behaviour in addToScene that cares about none.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -123,6 +123,56 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// restored to the pre-clip status afterwards.
+  static void clipRectAndPaint(Canvas canvas, Clip clipBehavior, Rect rect, Rect bounds, void painter()) {
+    _clipAndPaint(canvas, clipBehavior, rect, bounds, painter);
+  }
```
why are these on `Layer`? Seems like an odd place for them. Since they don't do anything with layers, i'd put them in the `painting` layer of our system.

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1141,6 +1149,16 @@ abstract class _RenderCustomClip<T> extends RenderProxyBox {
   T _clip;
 
+  /// {@template flutter.widget.clipper.clipBehavior}
+  /// Controls how to clip when clipper is null. Otherwise, clipper controls the
```
all identifiers of members in docs should be in `[]` every time (and all identifiers of arguments in docs should be in \`\` every time, but that doesn't apply here)

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1141,6 +1149,16 @@ abstract class _RenderCustomClip<T> extends RenderProxyBox {
+  /// {@endtemplate}
+  Clip _clipBehavior;
+
+  Clip get clipBehavior => clipper == null ? _clipBehavior : clipper.clipBehavior;
```
I think we should just expose the _clipBehavior directly, I don't see any reason to have a getter that defers to `clipper` sometimes.

liyuqian committed:
Address various comments

liyuqian committed:
Remove clipBehavior from CustomClipper

liyuqian committed:
Fix TestRecordingPaintingContext's clip behavior

liyuqian committed:
Nits

liyuqian commented:
@Hixie : updated. I just learned that Dart can implement non-abstract class and there's a noSuchMethod :smiley:

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577) modulo comments

we should put `ClipContext` in the `painting` layer, not `rendering`, since it's just about painting. I probably would have just made the `ClipContext` methods public global functions that took a `Canvas` as an argument actually (that'd be more like other things in the `painting` layer), but this is probably fine too. Might need rearranging at some point especially if we add more things to it.

liyuqian committed:
More fixes

1. add setter for clipBehavior to prevent Clip.none
2. move ClipContext to painting layer

liyuqian committed:
Fix import
